geomesa-raster/src/main/scala/org/locationtech/geomesa/raster/util/RasterUtils.scala

```diff
@@ -0,0 +1,291 @@
+    val colorModel = chunk.getColorModel
+    val alphaPremultiplied = colorModel.isAlphaPremultiplied
+    val sampleModel = chunk.getSampleModel.createCompatibleSampleModel(width, height)
+    val emptyRaster = JRaster.createWritableRaster(sampleModel, null)
+    new BufferedImage(colorModel, emptyRaster, alphaPremultiplied, properties)
+  }
+
+  def allocateBufferedImage(width: Int, height: Int, buf: BufferedImage): BufferedImage = {
+    val colorModel = buf.getColorModel
+    val alphaPremultiplied = colorModel.isAlphaPremultiplied
+    val emptyRaster = colorModel.createCompatibleWritableRaster(width, height)
+    new BufferedImage(colorModel, emptyRaster, alphaPremultiplied, null)
+  }
+
+  def renderedImageToBufferedImage(r: RenderedImage): BufferedImage = {
+    val properties = new JHashtable[String, Object]
+    if (r.getPropertyNames != null) {
+      r.getPropertyNames.foreach(name => properties.put(name, r.getProperty(name)))
+    }
+    val colorModel = r.getColorModel
+    val alphaPremultiplied = colorModel.isAlphaPremultiplied
+    val sampleModel = r.getSampleModel
+    new BufferedImage(colorModel, r.copyData(null), alphaPremultiplied, properties)
+  }
+
+  def writeToMosaic(mosaic: BufferedImage, raster: Raster, env: Envelope, resX: Double, resY: Double) = {
+    val croppedRaster = cropRaster(raster, env)
+    croppedRaster.foreach{ cropped =>
+        val rasterEnv = raster.referencedEnvelope.intersection(envelopeToReferencedEnvelope(env))
+        val originX = Math.floor((rasterEnv.getMinX - env.getMinimum(0)) / resX).toInt
+        val originY = Math.floor((env.getMaximum(1) - rasterEnv.getMaxY) / resY).toInt
+        mosaic.getRaster.setRect(originX, originY, cropped.getData)
+    }
+  }
+
+  def mosaicChunks(chunks: Iterator[Raster], queryWidth: Int, queryHeight: Int, queryEnv: Envelope): (BufferedImage, Int) = {
+    if (chunks.isEmpty) {
+      (null, 0)
+    } else {
+      val firstRaster = chunks.next()
+      if (!chunks.hasNext) {
+        val croppedRaster = cropRaster(firstRaster, queryEnv)
+        croppedRaster match {
+          case None      => (null, 1)
+          case Some(buf) => (scaleBufferedImage(queryWidth, queryHeight, buf), 1)
+        }
+      } else {
+        val accumuloRasterXRes = firstRaster.referencedEnvelope.getSpan(0) / firstRaster.chunk.getWidth
+        val accumuloRasterYRes = firstRaster.referencedEnvelope.getSpan(1) / firstRaster.chunk.getHeight
+        //TODO: check for corner cases: https://geomesa.atlassian.net/browse/GEOMESA-758
+        val mosaicX = Math.round(queryEnv.getSpan(0) / accumuloRasterXRes).toInt
+        val mosaicY = Math.round(queryEnv.getSpan(1) / accumuloRasterYRes).toInt
+        if (mosaicX <= 0 || mosaicY <= 0) {
+          (null, 1)
+        } else {
+          var count = 1
+          val mosaic = allocateBufferedImage(mosaicX, mosaicY, firstRaster.chunk)
+          writeToMosaic(mosaic, firstRaster, queryEnv, accumuloRasterXRes, accumuloRasterYRes)
```

Comment:
I don't think it should be done here thanks to the mutation being done to `mosaic`, but I think you could have done a Clever Thing here and wrapped the calls to `writeToMosaic` in a recursive function, which would have eliminated the var.